File: web/src/api/library-lookup.ts
import { basePath } from "@/lib/base-path";

export type LibraryLocationHit = {
  kind: "subscription" | "external";
  expand_key: string;
  title: string;
  enabled: boolean | null;
};

export type TrackPresence = {
  video_id: string;
  title: string | null;
  artist: string | null;
  in_direct: boolean;
  locations: LibraryLocationHit[];
};

export type PlaylistPresence = {
  url: string;
  subscription: LibraryLocationHit | null;
  in_direct_url: boolean;
};

export type TextMatchHit = {
  video_id: string;
  title: string;
  artist: string;
  in_direct: boolean;
  locations: LibraryLocationHit[];
};

export type TextPresence = {
  query: string;
  matches: TextMatchHit[];
};

async function lookupJson<T>(
  path: string,
): Promise<T | { error: string }> {
  const response = await fetch(`${basePath}/api${path}`, {
    credentials: "include",
  });
  if (!response.ok) {
    const body = (await response.json().catch(() => null)) as {
      detail?: string;
      message?: string;
    } | null;
    return {
      error:
        body?.detail ??
        body?.message ??
        `Lookup failed (${response.status})`,
    };
  }
  return (await response.json()) as T;
}

export function lookupTrackPresence(
  videoId: string,
): Promise<TrackPresence | { error: string }> {
  return lookupJson(
    `/library/lookup/track?video_id=${encodeURIComponent(videoId)}`,
  );
}

export function lookupPlaylistPresence(
  url: string,
): Promise<PlaylistPresence | { error: string }> {
  return lookupJson(
    `/library/lookup/playlist?url=${encodeURIComponent(url)}`,
  );
}

export function lookupTextPresence(
  query: string,
): Promise<TextPresence | { error: string }> {
  return lookupJson(
    `/library/lookup/text?q=${encodeURIComponent(query)}`,
  );
}

File: packages/api/src/yubal_api/schemas/library_lookup.py
"""Schemas for Enter-key local library quick lookup (operation B)."""

from __future__ import annotations

from typing import Literal

from pydantic import BaseModel, Field


class LibraryLocationHit(BaseModel):
    """One playlist / folder the track belongs to (excluding Direct)."""

    kind: Literal["subscription", "external"]
    expand_key: str
    title: str
    enabled: bool | None = None


class TrackPresenceResponse(BaseModel):
    video_id: str
    title: str | None = None
    artist: str | None = None
    in_direct: bool = False
    locations: list[LibraryLocationHit] = Field(default_factory=list)


class PlaylistPresenceResponse(BaseModel):
    url: str
    subscription: LibraryLocationHit | None = None
    in_direct_url: bool = False


class TextMatchHit(BaseModel):
    video_id: str
    title: str
    artist: str
    in_direct: bool = False
    locations: list[LibraryLocationHit] = Field(default_factory=list)


class TextPresenceResponse(BaseModel):
    query: str
    matches: list[TextMatchHit] = Field(default_factory=list)
